### DISCUSSION TITLE
Reporting feature for stock locations

### DISCUSSION THREAD

**Discussion by SergeoLacruz** [inventree/InvenTree#5110] [Ideas]
Hi,
The reporting system is a great thing for the documentation that is needed in some projects. Anyhow I recognized that there is no report feature on stock locations. Use case:
When we pack material for the factory we usually create a stock location for that box and put every stock item inside, physically and virtually. The box is sent to the factory. A report feature would allow me to automatically create a dispatch note (Lieferschein). 

Michael

**Comment by SchrodingersGat** [comment]
@SergeoLacruz reasonable idea, it would also be useful for a custom stocktake report for example. Is this a feature you would be looking to contribute?

**Comment by SergeoLacruz** [comment]
I would like to do so. But honestly I do not know where to start. any hints?

**Reply by SchrodingersGat** [reply]
Rundown of the required steps:

### Create new model

Create a new database model to define a new type of report - e.g. use SalesOrderReport as a starting point:

https://github.com/inventree/InvenTree/blob/752fb97d2f0da8e565635ec4e04a9ecb77e264b0/InvenTree/report/models.py#L476

### API Endpoints

You will need to create API endpoints for listing reports, and printing. 

Following the pattern for the SalesOrderReportMixin class, as follows:

https://github.com/inventree/InvenTree/blob/752fb97d2f0da8e565635ec4e04a9ecb77e264b0/InvenTree/report/api.py#L401

You will need to create 4x classes here:

- StockLocationReportMixin
- StockLocationReportList
- StockLocationReportDetail
- StockLocationReportPrint

And register the API URLs at the bottom of `api.py`

### Javasript Code

Final part is hooking into the report printing API in the front end. You can print reports either by selecting multiple stock locations (in a table) or from the "detail" page for a particular stock item.

Use the `printReports` javascript function, with the correct URLs:

https://github.com/inventree/InvenTree/blob/752fb97d2f0da8e565635ec4e04a9ecb77e264b0/InvenTree/order/templates/order/sales_order_base.html#L300

LMK how you go, happy to help further if you need

**Reply by SergeoLacruz** [reply]
Thanks for the hints. we will try this.

**Comment by wolflu05** [comment]
Just as cross reference, I think this is kinda related #2058

**Comment by xeno27** [comment]
I add a PR with this feature.
https://github.com/inventree/InvenTree/pull/5134

**Comment by SergeoLacruz** [comment]
Thanks to xeno27 for implementing this
